### PR TITLE
feat(ec2): support multi NIC/IP setups

### PR DESCRIPTION
**Pull request by aciba90**
Read individual commits for more context.

## Additional Context

https://warthogs.atlassian.net/browse/SC-1662
https://bootstack.canonical.com/cases/00336928
https://bootstack.canonical.com/cases/00377150

## Test Steps

Run the added integration tests on EC2.

## Checklist

- [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/development/index.html)
- [x] I have updated or added any [unit tests](https://cloudinit.readthedocs.io/en/latest/development/testing.html) accordingly
- [x] I have updated or added any [documentation](https://cloudinit.readthedocs.io/en/latest/development/contribute_docs.html) accordingly

## Merge type

- [ ] Squash merge using "Proposed Commit Message"
- [x] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

### CODE REVIEW

**Comment by rpocase** [comment]
I would be interested in @nmeyerhans 's feedback on this if you have some time. This is intended to provide multi-nic use cases for ubuntu based ec2 instances (i.e. a canonical supported alternative to ec2-net-utils). I don't know that this PR is intended to cover every use case yet.

**Review by TheRealFalcon — CHANGES_REQUESTED** [review]
Overall, looks good! I left some comments inline.

More broadly, do we need to use separate routing tables to accomplish this? Netplan/V2 allow specifying routes per interface. Can specifying `from` on a route rather than specifying a table work for this?

If not, this code is netplan-specific (not v2-specific), and we'll need to gate it as such.

File: cloudinit/config/cc_install_hotplug.py
Comment (nmeyerhans) [reply cloudinit/config/cc_install_hotplug.py:112]: These are the drivers that back ENIs in Amazon EC2.  This is the same list used by amazon-ec2-net-utils in Amazon Linux.

**Comment by nmeyerhans** [comment]
> More broadly, do we need to use separate routing tables to accomplish this? Netplan/V2 allow specifying routes per interface. Can specifying from on a route rather than specifying a table work for this?

The issue is that standard IP routing cannot take the source address of an outgoing packet into consideration.  Source-based routing (aka policy routing) is necessary because VPCs (in the default configuration in which src/dest checking is enabled) enforce that packets egressing an ENI must have a source address that matches the list of addresses or prefixes assigned to that ENI in the VPC control plane.  Packets that do not meet these criteria are dropped. Policy routing, which allows routing decisions based on additional factors such as the source address, is necessary in order to comply with VPCs restrictions.

A common scenario that triggers this relates to return traffic in response to an inbound packet to an instance with multiple attached ENIs.  That traffic will come in with a destination IP matching one of the instance's interfaces.  When the response is sent, that address will become the source address of the reply packet, and routing will be performed according to the entries in the main routing table.  The routing decision will take the _destination_ of the outgoing packet into account, but not the _source_.  Thus it won't necessarily result in this response packet egressing the same ENI on which the original packet was received.  If it egresses a different ENI, it will be dropped.  The policy rules let us ensure that the traffic is routed according to a table that only knows about the interfaces with a given source address associated with them in the VPC control plane.

There isn't a way to ensure routing compliance with these VPC restrictions with standard destination-based routing behavior.

File: cloudinit/sources/DataSourceEc2.py
Comment (nmeyerhans) [inline cloudinit/sources/DataSourceEc2.py:958]: One thing to be aware of when accessing IMDS in response to udev add events is that IMDS propagates asynchronously, and the data corresponding with a newly attached network interface may not be complete at the time that the handler is running.

Note also that device-number is not sufficient to uniquely identify an interface attachment.  Certain instance configuration may repeat device-number across multiple interfaces, for example as shown in https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/p5-efa.html

In those cases, you may also want to consult the `network-card` key, which represents the NetworkCardIndex API field.  It gets a bit confusing, though, because IMDS returns an http 404 error in multiple scenarios with different meanings:
* The interface attachment doesn't actually exist
* The interface attachment data hasn't yet propagated to IMDS (in which case you probably should retry)
* The returned value would be zero (which is a perfectly valid value and is seen on the instance's primary interface)

File: cloudinit/config/cc_install_hotplug.py
Comment (aciba90) [reply cloudinit/config/cc_install_hotplug.py:24]: Yes, we will need to rename that file at `preinst upgrade` time in the downstream ubuntu branches.

This is needed because `ENV{ID_NET_DRIVER}` is not defined when the rule is tested when defined in `10-cloud-init-hook-hotplug.rules`, so the rule does not match. I tried also with `DRIVERS` but the rule didn't match at NIC removal time.

File: tests/integration_tests/modules/test_hotplug.py
Comment (aciba90) [reply tests/integration_tests/modules/test_hotplug.py:154]: Thanks, we will refactor after that one is completed.

File: cloudinit/sources/DataSourceEc2.py
Comment (aciba90) [reply cloudinit/sources/DataSourceEc2.py:959]: The action of this key is to install any additional routes received from the DHCP server in the main routing table, and this is intentional, to be extra careful, and route packets not captured by the static routes, as for example, additional IPs added to secondary NICs.

This key is enabled by default by netplan when `dhcp==true`, see [1], thus no additional side-effects are performed, compared with previous behavior.

[1] https://netplan.readthedocs.io/en/0.106/netplan-yaml/#dhcp-overrides

File: tests/integration_tests/modules/test_hotplug.py
Comment (aciba90) [reply tests/integration_tests/modules/test_hotplug.py:215]: Done, thanks.

File: tests/integration_tests/test_networking.py
Comment (aciba90) [reply tests/integration_tests/test_networking.py:325]: Done, thanks.

File: cloudinit/sources/DataSourceEc2.py
Comment (aciba90) [reply cloudinit/sources/DataSourceEc2.py]: Each secondary NIC will use its own table. `nic_idx` had a double, one for computing the route metric and the other for the route table. I have refactored that bit, hopefully simplifying it.

File: cloudinit/sources/DataSourceEc2.py
Comment (aciba90) [reply cloudinit/sources/DataSourceEc2.py]: Thanks for pointing this out, I have added an extra commit handling the non-uniqueness of `device-index`.

**Comment by holmanb** [comment]
> As stated, I have gated the networking code to Netplan systems because `V2.routing-policy` only works if passed though the system.

Are we planning on fixing that in a follow-up PR? If not could we at least document that policy-based routing doesn't work on network v2 by filing a new issue and then referencing the issue number in this code?

At a minimum we should track deficiencies like that as we add features that depend on missing behavior - though ideally we would also update network v2 simultaneously so that this feature works everywhere.

File: tests/integration_tests/clouds.py
Comment (TheRealFalcon) [inline tests/integration_tests/clouds.py:215]: This functionality seems like it would be more appropriate in pycloudlib. I think it should also be addressed soon after this one lands.

File: tests/unittests/sources/test_ec2.py
Comment (TheRealFalcon) [inline tests/unittests/sources/test_ec2.py:1095]: Good set of tests for this function!

**Review by TheRealFalcon — APPROVED** [review]
Overall LGTM. I don't see anything major left, so I'm approving as-is, but I still have a few comments.

With this change, `convert_ec2_metadata_network_config` is getting a bit long and harder to follow. I think it'd be better if we could split the policy based routing pieces into its own function(s) and perhaps use a single implementation for the v4 and v6 pieces.

To echo Brett's comment, I think that in general it's a bad idea to add functionality specific to one renderer. We should ensure that our networking config is expressive enough to do what we need to do rather than work around it.

**Comment by holmanb** [comment]
@aciba90 Thanks for working on this. Cloud-init has historically not provided appropriate routes for non-primary interfaces - this is a step in the right direction towards a better multi-interface story.

I have a few things that I think we need to address:

1) I think we need to address @nmeyerhans's earlier comment:

> One thing to be aware of when accessing IMDS in response to udev add events is that IMDS propagates asynchronously, and the data corresponding with a newly attached network interface may not be complete at the time that the handler is running.

As it is, when the udev event triggers the hotplug_hook command, [cloud-init will `fetch()` the data once](https://github.com/canonical/cloud-init/blob/928e79029c0d22b629418db666eec7fa270653ad/cloudinit/cmd/devel/hotplug_hook.py#L188). As @nmeyerhans pointed out, the IMDS [might not have the interface attachment data](https://github.com/canonical/cloud-init/pull/4799#discussion_r1472287942). Won't this be a problem if it happens.

2) I'm glad that we're going to try to do the right thing automatically for pre-existing interfaces, but I wish we weren't going to force the user to opt in to hotplugged interfaces. It seems that you've figured out how, using `ENV{ID_NET_DRIVER}=="vif|ena|ixgbevf", GOTO="cloudinit_hook"`, to avoid the [docker issue that prompted opt-in in the first place](https://bugs.launchpad.net/cloud-init/+bug/1946003). Since that issue [was the whole reason for the creation of `cc_install_hotplug`](https://github.com/canonical/cloud-init/commit/1d01da5d9916d97ef463ba61a36b3f98f8911419), I really don't see why we couldn't just do the right thing for hotplugged interfaces. I know that in the design document you shared we discussed a cli argument to allow opting in, but really this shouldn't be necessary at all if we can install this hook by default.

3) While reviewing this code, I came across another issue that you didn't create, but I think is a real problem. Cloud-init's udev event handler script [completely ignores udev events while cloud-init is still running](https://github.com/canonical/cloud-init/blob/928e79029c0d22b629418db666eec7fa270653ad/tools/hook-hotplug#L11). We should fix this. It might be as simple as removing that check and then making `cloud-init-hotplugd.service` order `After=` some target. I would assume after `cloud-init.service` would probably be safe to make sure the config is merged but would have to test, cloud-final would be easier to order after, but might delay configuration longer than we have to.

**Review by holmanb — CHANGES_REQUESTED** [review]

File: cloudinit/config/cc_install_hotplug.py
Comment (aciba90) [reply cloudinit/config/cc_install_hotplug.py:24]: PR for noble: https://github.com/canonical/cloud-init/pull/4851, I will create similar ones for f, j, m.

**Comment by aciba90** [comment]
Thanks, @TheRealFalcon and @nmeyerhans, for the review and comments!
 
> Are we planning on fixing that in a follow-up PR? If not could we at least document that policy-based routing doesn't work on network v2 by filing a new issue and then referencing the issue number in this code?

> At a minimum we should track deficiencies like that as we add features that depend on missing behavior - though ideally we would also update network v2 simultaneously so that this feature works everywhere.

> To echo Brett's comment, I think that in general it's a bad idea to add functionality specific to one renderer. We should ensure that our networking config is expressive enough to do what we need to do rather than work around it.

Yes, I do agree. But being that potentially a more fundamental change, could it be a separate PR / ticket?

> With this change, convert_ec2_metadata_network_config is getting a bit long and harder to follow. I think it'd be better if we could split the policy based routing pieces into its own function(s) and perhaps use a single implementation for the v4 and v6 pieces.

Yeah, that's totally reasonable, let me create an issue to refactor that, I will handle it.

> 1. nmeyerhans's earlier comment:

>> One thing to be aware of when accessing IMDS in response to udev add events is that IMDS propagates asynchronously, and the data corresponding with a newly attached network interface may not be complete at the time that the handler is running.

> As it is, when the udev event triggers the hotplug_hook command, cloud-init will fetch() the data once. As @nmeyerhans pointed out, the IMDS might not have the interface attachment data. Won't this be a problem if it happens.

As I understood [nmeyerhans's comment](https://github.com/canonical/cloud-init/pull/4799#discussion_r1472287942), the data that could be missing is the `network-card` endpoint, not the rest of the metadata. I did not experience this race (`network-card` was present on the instance metadata fetched by cloud-init) while testing on `c6in.32xlarge`, `p4d.24xlarge`, `p4de.24xlarge`, `p5.48xlarge`. But even if we would experience the race, this won't be a big deal, because 

* cloud-init only needs to distinguish between primary and non primary NICs to add extra routing to the non-primary ones
* The primary NIC will always have device-number==0 and network-card==0, see https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_InstanceNetworkInterfaceSpecification.html .
* `DataSourceEc2._build_nic_order` will correctly sort the primary NIC even if network-card is not present.

> 2. Shouldn't it be possible to make udev work by default without opt-in?

> I'm glad that we're going to try to do the right thing automatically for pre-existing interfaces, but I wish we weren't going to force the user to opt in to hotplugged interfaces. It seems that you've figured out how, using ENV{ID_NET_DRIVER}=="vif|ena|ixgbevf", GOTO="cloudinit_hook", to avoid the docker issue that prompted opt-in in the first place. Since that issue was the whole reason for the creation of cc_install_hotplug, couldn't we just do the right thing for hot plugged interfaces - why require opt-in? 

I think that's a great idea but, we would need siblings allowed-lists of drivers for the clouds that support hotplug, otherwise we would be effectively reintroducing [LP: #1946003](https://bugs.launchpad.net/cloud-init/+bug/1946003) for other clouds. Thus, enabling it by default could be problematic.

> I know that in the design document you shared we discussed a cli argument to allow opt-in, but is that really necessary with this fix in place?

But, even if we would enable it by default, cloud-init does not make changes in an already boot instance in general, so the rules wouldn't be installed after an upgrade of cloud-init in a running instance, and the command is required if we want to provide that way in.


> 3. cloud-init's udev script is broken

> While reviewing this code, I came across another issue that you didn't create, but I think is a real problem. Cloud-init's udev event handler script completely ignores udev events while cloud-init is still running. This means that there is a race between installing the udev rule and completion where if a rule fires cloud-init won't configure the interface. Perhaps more concerning is that this also means that races are possible on all subsequent reboots.
We should fix this.

> It might be as simple as removing that check and then making cloud-init-hotplugd.service order After= some target. If that works to queue the arguments in the socket, it might be possible to queue after cloud-init.service or even earlier?

Yes, I noticed this racey behavior could happen too. Even if this seems trivial to implement, I'd like to split it as a different PR / unit of work if possible.

In general, I thought it would be positive to include this core functionality on 24.1 even if gated for Netplan/Ubuntu, to unblock some AWS teams requiring this feature. And, any extra hotplug related improvements/works could be done for the next release.

**Comment by TheRealFalcon** [comment]
> As it is, when the udev event triggers the hotplug_hook command, cloud-init will fetch() the data once. As nmeyerhans pointed out, the IMDS might not have the interface attachment data. Won't this be a problem if it happens.

@holmanb , our [hotplug hook](https://github.com/canonical/cloud-init/blob/main/cloudinit/cmd/devel/hotplug_hook.py/#L212) already has a fairly lengthy retry period. You linked to the code that fetchs the pickled datasource from disk.

> I think that's a great idea but, we would need siblings allowed-lists of drivers for the clouds that support hotplug, otherwise we would be effectively reintroducing [LP: #1946003](https://bugs.launchpad.net/cloud-init/+bug/1946003) for other clouds. Thus, enabling it by default could be problematic.

Not necessarily. If we add `EventType.HOTPLUG` to EC2's copy of [default_update_events](https://github.com/canonical/cloud-init/blob/main/cloudinit/sources/__init__.py/#L255), it should work unless overridden by userdata to be disabled. I was also initially hesitant to re-enable this automatically, but since this behavior already exists in Amazon Linux by default, I feel better about enabling it by default for us too.

> While reviewing this code, I came across another issue that you didn't create, but I think is a real problem. Cloud-init's udev event handler script [completely ignores udev events while cloud-init is still running](https://github.com/canonical/cloud-init/blob/928e79029c0d22b629418db666eec7fa270653ad/tools/hook-hotplug#L11). This means that there is a race between installing the udev rule and completion where if a rule fires cloud-init won't configure the interface. Perhaps more concerning is that this also means that races are possible on all subsequent reboots.
We should fix this.

Yes, but with some caveats. We don't want cloud-init hotplug handling early boot udev events because local timeframe has already handled those interfaces. At the very least, we need to wait until after network has come up to start handling udev events. Furthermore, we need to ensure that one of the normal boot stages of cloud-init isn't running when processing hotplug events as we don't want conflicts reading and writing the pickle. This shouldn't be hard to do, but just mentioning that it should be a consideration.

> In general, I thought it would be positive to include this core functionality on 24.1 even if gated for Netplan/Ubuntu, to unblock some AWS teams requiring this feature. And, any extra hotplug related improvements/works could be done for the next release.

Agreed. I think that most of these issues can be dealt with in follow-up PRs. However, I would like to see the tickets created before this PR merges with a 24.2 label attached, just so that we make sure we're tracking and prioritizing them accordingly.

File: cloudinit/config/cc_install_hotplug.py
Comment (aciba90) [inline cloudinit/config/cc_install_hotplug.py:62]: Just for the record, as this was a bit tricky to find.

[80-net-setup-link.rules](https://github.com/systemd/systemd/blob/main/rules.d/80-net-setup-link.rules) is a systemd-udev upstream rule which is installed by default that calls the [net_setup_setup builtin](https://github.com/systemd/systemd/blob/main/src/udev/udev-builtin-net_setup_link.c) and sets `ID_NET_DRIVER`.

One way to test it is:

```sh
$ SYSTEMD_LOG_LEVEL=debug udevadm test-builtin net_setup_link /sys/class/net/ens5
SELinux enabled state cached to: disabled
Trying to open "/etc/systemd/hwdb/hwdb.bin"...
Trying to open "/etc/udev/hwdb.bin"...
Trying to open "/usr/lib/systemd/hwdb/hwdb.bin"...
Trying to open "/lib/systemd/hwdb/hwdb.bin"...
Trying to open "/lib/udev/hwdb.bin"...
=== trie on-disk ===
tool version:      	249
file size:    	11126818 bytes
header size         	80 bytes
strings        	2374874 bytes
nodes          	8751864 bytes
Load module index
Found cgroup2 on /sys/fs/cgroup/, full unified hierarchy
Found container virtualization none.
Loaded timestamp for '/etc/systemd/network'.
Loaded timestamp for '/run/systemd/network'.
Parsed configuration file /usr/lib/systemd/network/99-default.link
Parsed configuration file /usr/lib/systemd/network/73-usb-net-by-mac.link
Parsed configuration file /run/systemd/network/10-netplan-ens5.link
Created link configuration context.
ID_NET_DRIVER=ena
... more stuff ...
```

Refs:

https://packages.fedoraproject.org/pkgs/systemd/systemd-udev/fedora-rawhide.html
https://packages.debian.org/search?mode=filename&suite=sid&section=all&arch=any&searchon=contents&keywords=80-net-setup-link.rules

**Review by aciba90 — COMMENTED** [review]
As agreed, I have enabled hotplug network updates for ec2 by default. PRs for gating this feature on Ubuntu downstream branches will come.

In addition to that, I did change the udev rule priority to 90- and documented why.

Please, @TheRealFalcon and @holmanb, re-review.

File: cloudinit/sources/DataSourceEc2.py
Comment (TheRealFalcon) [inline cloudinit/sources/DataSourceEc2.py:108]: Seeing it here, I'm not sure this combination makes sense. This essentially says "use the cached network config unless we get a hotplug event", but that means that if a network interface is added while the machine is powered off, we won't see it on a subsequent boot.

Not something to block this PR, but I think we need to consider adding `EventType.BOOT` here as well.

**Review by TheRealFalcon — APPROVED** [review]
LGTM! Thanks @aciba90 for the big effort here.

File: cloudinit/sources/DataSourceEc2.py
Comment (aciba90) [reply cloudinit/sources/DataSourceEc2.py:108]: Thanks for pointing this out and create https://bootstack.canonical.com/cases/00378972. This will be handled separately.

**Review by holmanb — APPROVED** [review]
Thanks for working on this @aciba90. Looks good to me!

**Comment by xdxu** [comment]
This PR essentially broke the routing logic in our production code. We will have to clean up these redundant PBR rules to get things back to work.
IMHO, how we route the traffic (either local out or pass through traffic) in a multi-homed instance should be up to the end user, rather than relying on these automatically generated rules.
My comment is based on the assumption that Linux instances running on AWS EC2 is for generic purposes and we can't really assume that most users would like this new behavior. There are a lot of ways to implement multi-wan like routing scheme and most of them will be taking advantage of the PBR rules, which means these extra rules will likely break those systems.

**Comment by aciba90** [comment]
@xdxu sorry about the breaking changes. We thought that providing a default working environment would add more value and that advanced users / use-cases could [disable cloud-init's network configuration](https://docs.cloud-init.io/en/latest/reference/network-config.html#disabling-network-configuration) or provide the custom network configuration via cloud-init or other means.

If you still disagree, please open a new issue to discuss it with more visibility, or consider hanging out in #cloud-init [IRC channel on Libera](https://kiwiirc.com/nextclient/irc.libera.chat/cloud-init).

Thanks.

**Comment by xdxu** [comment]
Thanks for the reply @aciba90
I created an issue https://github.com/canonical/cloud-init/issues/5249 for further discussion.

**Pull request merged**
